doc/devguide/extending/app-layer/transactions.rst

```diff
@@ -0,0 +1,321 @@
+- Simpler scenarios: state is simply an int.  ``1`` represents transaction completion, per direction.
+- Complex Transaction State in Suricata: ``enum`` (Rust: ``i32``). Completion is indicated by the highest enum value (some examples are: SSH, HTTP, HTTP2, DNS, SMB).
+
+Examples
+========
+
+This section shares some examples from Suricata codebase, to help visualize how Transactions work and are handled by the engine.
+
+Enums
+~~~~~
+
+Code snippet from: rust/src/ssh/ssh.rs:
+
+.. code-block:: rust
+
+    pub enum SSHConnectionState {
+        SshStateInProgress = 0,
+        SshStateBannerWaitEol = 1,
+        SshStateBannerDone = 2,
+        SshStateFinished = 3,
+    }
+
+From src/app-layer-ftp.h:
+
+.. code-block:: c
+
+    enum {
+        FTP_STATE_IN_PROGRESS,
+        FTP_STATE_PORT_DONE,
+        FTP_STATE_FINISHED,
+    };
+
+From src/app-layer-ssl.h:
+
+.. code-block:: c
+
+    enum {
+        TLS_STATE_IN_PROGRESS = 0,
+        TLS_STATE_CERT_READY = 1,
+        TLS_HANDSHAKE_DONE = 2,
+        TLS_STATE_FINISHED = 3
+    };
+
+API Callbacks
+~~~~~~~~~~~~~
+
+In Rust, this is done via the RustParser struct. As seen in rust/src/applayer.rs:
+
+.. code-block:: rust
+
+    /// Rust parser declaration
+    pub struct RustParser {
+            .
+            .
+            .
+        /// Progress values at which the tx is considered complete in a direction
+        pub tx_comp_st_ts:      c_int,
+        pub tx_comp_st_tc:      c_int,
+        .
+        .
+        .
+    }
+
+In C, the callback API is:
+
+.. code-block:: c
+
+    void AppLayerParserRegisterStateProgressCompletionStatus(
+        AppProto alproto, const int ts, const int tc)
+
+Simple scenario described, in Rust:
+
+rust/src/dhcp/dhcp.rs:
+
+.. code-block:: rust
+
+    tx_comp_st_ts: 1
+    tx_comp_st_tc: 1
+
+For SSH, this looks like this:
+
+rust/src/ssh/ssh.rs:
+
+.. code-block:: rust
+
+    tx_comp_st_ts: SSHConnectionState::SshStateFinished as i32,
+    tx_comp_st_tc: SSHConnectionState::SshStateFinished as i32,
+
+In C, callback usage would be as follows:
+
+src/app-layer-dcerpc.c:
+
+.. code-block:: c
+
+    AppLayerParserRegisterStateProgressCompletionStatus(ALPROTO_DCERPC, 1, 1);
+
+src/app-layer-ftp.c:
+
+.. code-block:: c
+
+    AppLayerParserRegisterStateProgressCompletionStatus(
+        ALPROTO_FTP, FTP_STATE_FINISHED, FTP_STATE_FINISHED);
+
+Sequence Diagrams
+~~~~~~~~~~~~~~~~~
+
+A DNS transaction in Suricata can be considered unidirectional:
+
+.. image:: diagrams/DnsUnidirectionalTransactions.png
+  :width: 600
+  :alt: A sequence diagram with two entities, Client and Server, with an arrow going from the Client to the Server, labeled "DNS Request". After that, there is a dotted line labeled "Transaction Completed".
+
+An HTTP2 transaction is an example of a bidirectional transaction, in Suricata (note that transactions in HTTP2 may
+overlap, scenario not shown in this Sequence Diagram):
+
+.. TODO add another example for overlapping HTTP2 transaction
```

Comment:
the txs are mapped to http2 "streams". Multiple streams can run in parallel, but don't interfere with each other